qualtran/bloqs/gf_arithmetic/gf2_inverse.py

```diff
@@ -0,0 +1,167 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import Dict, Set, TYPE_CHECKING, Union
+
+import attrs
+import numpy as np
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BloqDocSpec,
+    DecomposeTypeError,
+    QGF,
+    Register,
+    Side,
+    Signature,
+)
+from qualtran.bloqs.gf_arithmetic.gf2_addition import GF2Addition
+from qualtran.bloqs.gf_arithmetic.gf2_multiplication import GF2Multiplication
+from qualtran.bloqs.gf_arithmetic.gf2_square import GF2Square
+from qualtran.symbolics import is_symbolic, SymbolicInt
+
+if TYPE_CHECKING:
+    from qualtran import BloqBuilder, Soquet
+    from qualtran.resource_counting import BloqCountDictT, BloqCountT, CostKey, SympySymbolAllocator
+    from qualtran.simulation.classical_sim import ClassicalValT
+
+
+@attrs.frozen
+class GF2Inverse(Bloq):
+    r"""Out of place inversion for elements in GF($2^m$)
+
+    Given a quantum register storing elements from GF($2^m$), this bloq computes the inverse
+    of the given element in a new output register, out-of-place. Specifically,
+    it implements the transformation
+
+    $$
+        |a\rangle |0\rangle \rightarrow |a\rangle |a^{-1}\rangle
+    $$
+
+    Inverse is computed by using Fermat's little theorem for Finite Fields, which states that
+    for a finite field $\mathbb{F}$ with $m$ elements, $\forall a \in \mathbb{F}$
+    $$
+        a^{m} = a
+    $$
+
+    When the finite field is GF($2^m$), Fermat's little theorem can be used to obtain the relation
+
+    $$
+        a^{-1} = a^{2^m - 2}
+    $$
+
+    Thus, the inverse can be obtained via $m - 1$ squaring and multiplication operations.
+
+    Args:
+        bitsize: The degree $m$ of the galois field $GF(2^m)$. Also corresponds to the number of
+            qubits in the input register whose inverse should be calculated.
+
+    Registers:
```

Comment:
is there a reference for this construction?

Reply:
This one follows directly from Fermat's little theorem and the standard repeated squaring for exponentiation, that I've explained in the docstring. Dmitri told me there's a more optimized version explain in https://www.sciencedirect.com/science/article/pii/0890540188900247, which reduces the cost from O(m^3) to O(m^2log(m)) via a smarter way to implement repeated squaring for multiplication. 

I'll implement this in a follow-up and add a reference.